diff --git a/arch/arm/cpu/armv7/start.S b/arch/arm/cpu/armv7/start.S
index 29694b3b3..9e05514a2 100644
--- a/arch/arm/cpu/armv7/start.S
+++ b/arch/arm/cpu/armv7/start.S
@@ -221,12 +221,18 @@ skip_errata_454179:
 skip_errata_454179:
 #endif
 
-#ifdef CONFIG_ARM_ERRATA_430973
+#if defined(CONFIG_ARM_ERRATA_430973) || defined(CONFIG_POCKETCHIP_CORTEX_A8_SPECTRE_V2)
+#ifndef CONFIG_POCKETCHIP_CORTEX_A8_SPECTRE_V2
 	cmp	r2, #0x21		@ Only on < r2p1
 	bge	skip_errata_430973
+#endif
 
 	mrc	p15, 0, r0, c1, c0, 1	@ Read ACR
+#ifdef CONFIG_POCKETCHIP_CORTEX_A8_SPECTRE_V2
+	orr	r0, r0, #(0x1 << 6)	@ Set IBE bit always for Cortex-A8 Spectre v2
+#else
 	orr	r0, r0, #(0x1 << 6)	@ Set IBE bit
+#endif
 	push	{r1-r5}			@ Save the cpu info registers
 	bl	v7_arch_cp15_set_acr
 	pop	{r1-r5}			@ Restore the cpu info - fall through
